inloop/statistics/views.py

```diff
@@ -0,0 +1,83 @@
+import json
+from json import JSONDecodeError
+
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.db import models
+from django.http import Http404, JsonResponse
+from django.views import View
+from django.views.generic import TemplateView
+
+from inloop.solutions.models import Solution
+from inloop.statistics.truncators import GranularityUnavailable, get_truncator
+from inloop.tasks.models import Category
+
+from .responses import IncorrectAccessMethod, MalformedJson
+
+
+class AdminView(UserPassesTestMixin, LoginRequiredMixin, View):
+    def test_func(self):
+        if not self.request.user:
+            return False
+        return self.request.user.is_superuser or self.request.user.is_staff
+
+
+class StatisticsView(AdminView, TemplateView):
+    template_name = "statistics/index.html"
+
+    def get_context_data(self):
+        return {
+            "solutions_histogram": {
+                "available_categories": [
+                    {
+                        "name": category.name,
+                        "id": category.id
+                    } for category in Category.objects.all()
+                ]
+            }
+        }
+
+
+class SolutionsHistogramView(AdminView, TemplateView):
+    template_name = "statistics/solutions_histogram.html"
+
+    def post(self, request) -> JsonResponse:
+        if not request.user:
+            raise Http404()
+
+        if not request.user.is_staff and not request.user.is_superuser:
+            raise Http404()
+
+        try:
+            data = json.loads(request.body)
+        except JSONDecodeError:
+            return MalformedJson()
+
+        from_timestamp = data.get("from_timestamp", None)
+        to_timestamp = data.get("to_timestamp", None)
+        passed = data.get("passed", None)
+        category_id = data.get("category_id", None)
+        granularity = data.get("granularity", "minute")
+
+        try:
+            truncator = get_truncator(field_name="submission_date", granularity=granularity)
+        except GranularityUnavailable:
+            return MalformedJson()
+
+        queryset = Solution.objects
+        if from_timestamp:
+            queryset = queryset.filter(submission_date__gte=from_timestamp)
+        if to_timestamp:
+            queryset = queryset.filter(submission_date__lte=to_timestamp)
+        if passed is not None:
+            queryset = queryset.filter(passed=passed)
+        if category_id:
+            queryset = queryset.filter(task__category_id=category_id)
+
+        histogram_queryset = queryset \
+            .annotate(date=truncator) \
+            .values("date") \
+            .annotate(count=models.Count("pk"))
+
+        return JsonResponse({
+            "histogram": list(histogram_queryset)
```

Comment:
This will serialize the dates to strings, right?

Reply:
As far as I remember (this code was implemented some weeks ago) this is currently serializing the dates into strings. In the couse of cleaning up, it would of course be possible to replace this with a list comprehension that converts the dates into unix timestamps, to improve performance.